Dist actor resolve silgen

kavon opened the pull request.

kavon committed:
[distributed] a number of ctor infra cleanups

1. We're no longer synthesizing the local
init. All designated inits of a dist actor
must have exactly one ActorTransport parameter.
This parameter is used in the init's prologue
to initialize some of its members.

2. We're synthesizing a factory function for
the resolve initialization, instead of an
actor-member named `init` to serve that
purpose.

As such, much of the earlier infrastructure
is no longer needed, etc.

kavon committed:
[distributed] synthesize factory resolve func

ktoso commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -1,8 +1,8 @@
-//===--- CodeSynthesis.cpp - Type Checking for Declarations ---------------===//
+//===--- CodeSynthesisDistributedActor.cpp --------------------------------===//
```
whoops thanks

ktoso commented on stdlib/public/Distributed/DistributedActor.swift:
```diff
@@ -36,34 +36,6 @@ public protocol AnyActor: AnyObject {}
-
-    @available(*, deprecated, renamed: "SomeDistributedActor.resolve(_:using:)")
-    init(resolve id: AnyActorIdentity, using transport: ActorTransport) throws
-
```
@kavon okey so the reason the conformance issue you saw yesterday is this:

- since we removed the inits from the protocol
- we won't run TypeChecker::addImplicitConstructors
- or at least we won't get to invoke inside it it seems `addImplicitDistributedActorMembersToClass`
- this entry point used to be for the inits, but as we now it moved on and actually synthesizes more and not just the inits

So removing the req, causes the synthesis to not trigger.

You can for now keep the silly init(transport: ActorTransport) in this PR and you should be unblocked.
I'm in the meantime working on fixing that in my local initializer rework.

kavon committed:
Generalize BodyKind::MemberwiseInitializer

We'd like to support factor initializers
for distributed actor types that are
synthesized by SILGen.

We already do something similar for
memberwise initializers for structs.
Thus, this patch generalizes that
concept into a new BodyKind for
AbstractFunctionDecls called
BodyKind::SILSynthesize.

In addition, to help differentiate the
kinds of AFDs that are SILSynthesized
into different families for SILGen to
recognize, we also have a new enum
SILSynthesizeKind to indicate whether it
is a memberwise init, etc.

kavon committed:
[distributed] working on runtime allocation function

kavon commented on lib/SILGen/SILGenConstructor.cpp:
```diff
@@ -1211,3 +1187,30 @@ void SILGenFunction::emitIVarInitializer(SILDeclRef ivarInitializer) {
+  auto builtinName = ctx.getIdentifier(
+      getBuiltinName(BuiltinValueKind::InitializeDistributedRemoteActor));
+  auto returnType = getLoweredType(actor->getInterfaceType());
+  auto *metaTypeInfo = F.getArgument(0);
```
Probably the wrong argument number.

ktoso replied:
That'll be the last param, I'll handle that :)

kavon commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -1752,10 +1752,20 @@ void swift::swift_defaultActor_deallocateResilient(HeapObject *actor) {
-  auto actor = asImpl(_actor);
-  actor->initialize(/*remote=*/true);
+// TODO: rename initialize to allocate throughout htis built-in. Delete `swift_distributedActor_remote_create` throughout.
+OpaqueValue* swift::swift_distributedActor_remote_initialize(Metadata *actorType) {
```
>  error: out-of-line definition of 'swift_distributedActor_remote_initialize' does not match any declaration in namespace 'swift'

Didn't have time to dig into why.

ktoso replied:
It has a definition in  a header file as well, there it was still DefaultActor, easily fixed :)

ktoso replied:
Okey so this is the "create", if we are to follow concurrency lib convetions, where "initialize" used to be "there is some memory, initialize _into it_", like the actor does. Impl looks nice though - very excited if this is enough to make it work ❤️

ktoso commented on lib/SILGen/SILGenConstructor.cpp:
```diff
@@ -1211,3 +1187,30 @@ void SILGenFunction::emitIVarInitializer(SILDeclRef ivarInitializer) {
+  ///       demanded. For example, by declaring the actor as `public` or
+  ///       having at least one call to the resolve init.
+
+  SILLocation loc = fd; // NOTE: forgot if this is the right one for all locs.
```
I think it's good enough hm hm

ktoso commented:
Pulled into https://github.com/apple/swift/pull/38823 as cherry picked commits, we're almost there :)
